substrate/primitives/time/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! Substrate time and duration types.
+//!
+//! Main design goal of this API is to KISS (Keep It Simple and Stupid) and to still fulfill our
+//! needs. This means to rely on already existing conventions as much as possible.
+
+use codec::{EncodeLike, FullCodec};
+use core::fmt;
+use sp_arithmetic::traits::{CheckedAdd, CheckedSub, Saturating, Zero};
+use sp_runtime::traits::Member;
+
+/// Provides the current time.
+pub trait InstantProvider<I: Instant> {
+	/// Monotonicity assumptions about `now()`.
+	///
+	/// Some pallets may restrict themselves to only work with monotonic clocks.
+	const MONOTIC: Monotonicity;
+
+	/// Returns the current time.
+	fn now() -> I;
+}
+
+/// What to assume about two directly consecutive calls to a function `f`.
+pub enum Monotonicity {
+	/// No implied correlation between two calls to `f`.
+	None,
+	/// `f() <= f()` always holds.
+	MonotonicIncrease,
+	/// `f() < f()` always holds.
+	StrictMonotonicIncrease,
+}
+
+/// An instant or "time point".
+pub trait Instant:
+	Member
+	+ FullCodec
+	+ EncodeLike
+	+ fmt::Debug
+	+ scale_info::TypeInfo
+	+ core::cmp::PartialOrd
+	+ core::cmp::Eq
+	+ Since<<Self as Instant>::Delta>
+	+ Until<<Self as Instant>::Delta>
+	// If it turns out that our API is bad, then people can still use UNIX formats:
+	+ TryFrom<UnixInstant>
+	+ TryInto<UnixInstant>
+{
+	type Delta: Duration;
+
+	/// Dial time forward by `delta`.
+	fn checked_forward(&self, delta: &Self::Delta) -> Option<Self>;
+
+	/// Dial time backward by `delta`.
+	fn checked_rewind(&self, delta: &Self::Delta) -> Option<Self>;
+}
+
+/// A duration or "time interval".
+///
+/// Durations MUST always be positive.
+pub trait Duration:
+	Member
+	+ FullCodec
+	+ EncodeLike
+	+ fmt::Debug
+	+ scale_info::TypeInfo
+	+ core::cmp::PartialOrd
+	+ core::cmp::Eq
+	+ CheckedAdd
+	+ CheckedSub
+	+ Saturating
+	+ Zero
+	+ TryFrom<UnixDuration>
+	+ TryInto<UnixDuration>
+{
+	/// Scale the duration by a factor.
+	fn checked_scale(&self, other: u32) -> Option<Self>;
+}
+
+/// Calculates the time since a given instant.
+pub trait Since<D: Duration> {
```

Comment:
> Expecting this to be highly WIP 

Yes thanks for the feedback 😄  

I would also rather like to use `CheckedAdd` and `CheckedSub`, but they can only operate on `(self, self)` and not `(self, Duration)`. So I think that we need new traits for this.  

The Add trait can specify the right-hand operand, but the `Checked*` traits somehow dont: https://docs.rs/num/latest/num/trait.CheckedAdd.html

Reply:
yeah kinda sucks that they don't. 

what about `Instant` and `Duration` being the same type/trait? Can you give some examples of why there needs to be two traits for these two concepts?

Reply:
I put some points down here: https://github.com/paritytech/polkadot-sdk/pull/3298#issuecomment-1962992686  
I think if we use the same type, then we can only ever have `Duration`. Since what should it return when we subtract two `Instant`s?  
And that duration would need to be anchored somewhere, maybe at genesis time. But then we only ever have relative time and not absolute. It seems a bit weird to me, and the fact that Rust, C++ and other major ecosystems all split it into two types make me think that they are onto somethine.